backend-rust/src/indexer.rs

```diff
@@ -1919,10 +2132,181 @@ impl PreparedContractUpdate {
         .execute(tx.as_mut())
         .await?;
 
-        Ok(())
+        process_cis2_events(
+            &self.cis2_events,
+            self.contract_index,
+            self.contract_sub_index,
+            pool,
+            transaction_index,
+            tx,
+        )
+        .await
     }
 }
 
+async fn process_cis2_events(
+    cis2_events: &[cis2::Event],
+    contract_index: i64,
+    contract_sub_index: i64,
+    pool: &PgPool,
+    transaction_index: i64,
+    tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+) -> anyhow::Result<()> {
+    for log in cis2_events.iter() {
+        // The `total_supply` value of a token is inserted/updated in the database here.
+        // Only `Mint` and `Burn` events affect the `total_supply` of a
+        // token.
+        if let cis2::Event::Mint {
+            token_id,
+            amount,
+            owner,
+        } = log
+        {
+            let token_address =
+                get_token_address(contract_index as u64, contract_sub_index as u64, token_id)?;
+
+            // Fetch the current `total_supply` for the given `token_address`.
+            let row = sqlx::query!(
+                "
+                SELECT total_supply FROM tokens WHERE token_address = $1",
+                token_address,
+            )
+            .fetch_optional(pool)
+            .await?;
+
+            // If current `total_supply` exists, add the `amount` to it.
+            // otherwise use the `amount` as the new `total_supply`.
+            // Note: Some `buggy` CIS2 token contracts might mint more tokens than the
+            // MAX::TOKEN_AMOUNT specified in the CIS2 standard. The
+            // `total_supply` eventually overflows in that case.
+            let new_total_supply = if let Some(row) = row {
+                let current_total_supply: BigDecimal = row.total_supply;
+
+                current_total_supply + BigDecimal::from_str(&amount.0.to_string())?
+            } else {
+                BigDecimal::from_str(&amount.0.to_string())?
```

Comment:
Use [BigDecimal::from_biguint](https://docs.rs/bigdecimal/latest/bigdecimal/struct.BigDecimal.html#method.from_biguint) instead of converting to a string